src/python/pants/backend/python/rules/pytest_runner.py

```diff
@@ -228,26 +240,47 @@ async def run_python_test(
     test_options: TestOptions,
 ) -> TestResult:
     """Runs pytest for one target."""
-    env = {"PYTEST_ADDOPTS": f"--color={'yes' if global_options.options.colors else 'no'}"}
+    add_opts = [f"--color={'yes' if global_options.options.colors else 'no'}"]
+    test_results_file = f"{config.address.path_safe_spec}.xml"
+    if test_setup.xml_dir:
+        add_opts.extend(
+            (f"--junitxml={test_results_file}", f"-o junit_family={test_setup.junit_family}",)
+        )
+    env = {"PYTEST_ADDOPTS": " ".join(add_opts)}
+
     run_coverage = test_options.values.run_coverage
+    output_dirs = [".coverage"] if run_coverage else []
+    if test_setup.xml_dir:
+        output_dirs.append(test_results_file)
     process = test_setup.test_runner_pex.create_process(
         python_setup=python_setup,
         subprocess_encoding_environment=subprocess_encoding_environment,
         pex_path=f"./{test_setup.test_runner_pex.output_filename}",
         pex_args=test_setup.args,
         input_files=test_setup.input_files_digest,
-        output_directories=(".coverage",) if run_coverage else None,
+        output_directories=tuple(output_dirs) if output_dirs else None,
         description=f"Run Pytest for {config.address.reference()}",
         timeout_seconds=(
             test_setup.timeout_seconds if test_setup.timeout_seconds is not None else 9999
         ),
         env=env,
     )
     result = await Get[FallibleProcessResult](Process, process)
-    coverage_data = (
-        PytestCoverageData(config.address, result.output_directory_digest) if run_coverage else None
+    output_digest = result.output_directory_digest
+
+    coverage_data = PytestCoverageData(config.address, output_digest) if run_coverage else None
+    xml_test_results_digest: Optional[Digest] = None
+    if test_setup.xml_dir:
+        snapshot = await Get[Snapshot](
```

Comment:
It'd be good to call this `xml_snapshot` to make it more clear that this isn't as generic as `snapshot` implies.